src/mode/auth.rs

```diff
@@ -0,0 +1,119 @@
+
+#[derive(Parser, Debug)]
+pub(crate) enum AuthArgs {
+    Login {
+        /// contains destination of auth server
+        #[clap(short, long)]
+        provider: Option<String>,
+    },
+    Logout {
+        /// contains destination of auth server
+        #[clap(short, long)]
+        provider: Option<String>,
+    },
+}
+
+#[derive(Parser, Debug)]
+pub struct Args {
+    #[command(subcommand)]
+    auth: AuthArgs,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CliLoginResponse {
+    pub login_id: String,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CliLogin {
+    pub app_id: String,
+    // pub user_id: String,
+}
+
+static CTRL_URL: &str = "http://localhost:3000";
+
+pub(crate) async fn start(args: Args) -> Result<()> {
+    let mut config_manager = ConfigManager::new().unwrap();
+    match args.auth {
+        AuthArgs::Logout { provider } => {
+            let base_url = provider.unwrap_or_else(|| CTRL_URL.to_string());
+            config_manager.logout_provider(base_url);
+        }
+        AuthArgs::Login { provider } => {
+            let base_url = provider.unwrap_or_else(|| CTRL_URL.to_string());
+            let client = reqwest::Client::new();
+            let res = client
+                .post(format!("{base_url}/api/cli/login"))
+                .json(&CliLogin {
+                    app_id: config_manager.get_app_id(),
+                })
+                .send()
+                .await
+                .expect("failed to send cli/login request")
+                .json::<CliLoginResponse>()
+                .await
+                .expect("failed to parse JSON from cli/login");
+
+            let login_id =
+                url::form_urlencoded::byte_serialize(res.login_id.as_bytes()).collect::<String>();
+            let app_id =
+                url::form_urlencoded::byte_serialize(config_manager.get_app_id().as_bytes())
+                    .collect::<String>();
+            info!("\n\nPlease visit the following URL to authenticate this cli app {base_url}/cli/authenticate/{app_id}?login_id={login_id}\n\n");
+
+            let status_url = format!("{base_url}/api/cli/login/{}", res.login_id);
+            let auth_status = check_auth_status(&status_url, &client).await;
+            if auth_status.is_empty() {
+                panic!("Cli Login failed");
+            }
+
+            config_manager.add_provider(Provider {
+                name: base_url,
+                cookies: auth_status,
+            });
+        }
+    }
+
+    config_manager.flush()?;
+    Ok(())
+}
+
+async fn check_auth_status(status_url: &str, client: &reqwest::Client) -> Vec<String> {
+    loop {
+        match client.get(status_url).send().await {
+            Ok(res) => {
+                if res.status() == StatusCode::OK {
+                    return res
+                        .cookies()
+                        .map(|cookie| cookie.value().to_string())
+                        .collect();
```

Comment:
Hmm think we should store the whole cookie not just its `value` which is `session_id`.
See [CookieStore::save_json](https://docs.rs/reqwest_cookie_store/0.5.0/reqwest_cookie_store/struct.CookieStore.html#method.save_json) I always get 401 from the api since the cookie is not set properly.

`DEBUG cookie_store::cookie_store] unable to store Set-Cookie: Expired`
`"set-cookie": "lunatic_auth=; Path=/; expires=Thu, 01 Jan 1970 00:00:00 GMT"} }`